---
title: "An Introduction To ngsReports"
author:
- name: Christopher Ward
  affiliation: School of Biological Sciences, University of Adelaide
- name: Hien To
  affiliation: Bioinformatics Hub, University of Adelaide
- name: Steve Pederson
  affiliation: Bioinformatics Hub, University of Adelaide
  email: stephen.pederson@adelaide.edu.au
package: ngsReports
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{An Introduction To ngsReports}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```


# Introduction

The package `ngsReports` is designed to bolt into data processing pipelines and 
produce combined plots for FastQC reports across an entire set of libraries.
Output as generated by the tool 
[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) 
will be loaded and replotted using functions from `r CRANpkg("ggplot2")` and 
`r CRANpkg("plotly")`, as well as custom summary plots.
Tables of read counts can also be easily generated.

# Basic Usage

## Using the Shiny App

In addition to the usage demonstrated below, a `shiny` app has been developed 
for interactive viewing of FastQC reports.
This can be installed using:

```
remotes::install_github(UofABioinformaticsHub/fastqcRshiny)
```

A vignette for this app will be installed with the package.

## The default report

In it's simplest form, a default report can be generated simply by specifying 
a directory containing the output from FastQC and calling the function 
`writeHtmlReport()`.

```{r}
library(ngsReports)
```

```
fileDir <- file.path("path", "to", "your", "FastQC", "Reports")
writeHtmlReport(fileDir)
```

This function will transfer the default template to the provided directory and 
produce a single `.html` file containing interactive summary plots of any 
FastQC output found in the directory.
FastQC output can be `*fastqc.zip` files or the same files extracted as 
individual directories.

The default template is provided as `ngsReports_Fastqc.Rmd` in the package 
directory 
(`system.file("extdata", "ngsReports_Fastqc.Rmd", package = "ngsReports")`).
This can be easily modified and supplied as an alternate template to the above 
function using your modified file as the template RMarkdown file.

```
altTemplate <- file.path("path", "to", "your", "new", "template.Rmd")
writeHtmlReport(fileDir, template = altTemplate)
```

# Advanced Usage

## Classes Defined in the Package

The package `ngsReports` introduces four basic `S4` classes: 

* `FastqcFile` & `FastqcFileList`
* `FastqcData` & `FastqcDataList`

`FastqcFile` and `FastqcData` objects respectively hold the location and data 
from a single report as generated by the stand-alone tool 
`FastQC`^[https://www.bioinformatics.babraham.ac.uk/projects/fastqc/].
These are then extended into lists for more than one file.
`FastqcFile` and `FastqcFileList` objects simply contain the path to these 
reports after checking the files contain the files always output by this tool.
`FastqcData` and `FastqcDataList` objects contain the *parsed data* from the 
reports.
For most users, the primary class of interest will be the `FastqcDataList` and 
the function `getFastqcData()`

## Loading FastQC Data Into `R`

To load a set of `FastQC` reports into `R` as a `FastqcDataList`, specify the set of files, then call the function `getFastqcData()`.

```{r}
fileDir <- system.file("extdata", package = "ngsReports")
files <- list.files(fileDir, pattern = "fastqc.zip$", full.names = TRUE)
fdl <- getFastqcData(files)
```

From here, all FastQC modules can be obtained as a `tibble` (i.e. `data.frame`) 
using the function `getModule()` and choosing one of the following modules:

* `Summary` (The PASS/WARN/FAIL status for each module)
* `Basic_Statistics`
* `Per_base_sequence_quality`
* `Per_sequence_quality_scores`
* `Per_base_sequence_content`
* `Per_sequence_GC_content`
* `Per_base_N_content`
* `Sequence_Length_Distribution`
* `Sequence_Duplication_Levels`
* `Overrepresented_sequences`
* `Adapter_Content`
* `Kmer_Content`
* `Per_tile_sequence_quality`

```{r}
getModule(fdl[[1]], "Summary")
```

Capitalisation and spelling of these module names follows the default patterns 
from FastQC reports with spaces replaced by underscores.
One additional module is available and taken directly from the text within the 
supplied reports

* `Total_Duplicated_Percentage`

In addition, the read totals for each file in the library can be obtained using `readTotals()`, which can be easily used to make a table of read totals.
This essentially just returns the first two columns from `getModule(x, "Basic_Statistics")`

```{r, results='hide'}
reads <- readTotals(fdl)
```

The packages `dplyr` and `pander` can also be extremely useful for manipulating 
and displaying imported data.
To show only the R1 read totals, you could do the following

```{r}
library(dplyr)
library(pander)
dplyr::filter(reads, grepl("R1", Filename)) %>% 
  pander(
      big.mark = ",",
      caption = "Read totals from R1 libraries", 
      justify = "lr")
```


## Generating Plots

Plots created from a single `FastqcData` object will resemble those generated 
by the `FastQC` tool, whilst those created from a `FastqcDataList` will be 
combined summaries across a library of files.
All plots are able to be generated as interactive plots using the argument 
`usePlotly = TRUE`.

Most modules have been enabled for plotting using default `S4` dispatch.

### plotSummary()

A summary of all `PASS/WARN/FAIL` flags can be simply generated

```{r plotSummary, fig.cap="Default summary of FastQC flags.", fig.wide = TRUE}
plotSummary(fdl)
```

### plotReadTotals()

Read totals from all files in the library can be simply plotted.
By default, the number of duplicated sequences from the 
`Total_Duplicated_Percentage` module are shown, but this can be disabled by 
setting `duplicated = FALSE`.

```{r}
plotReadTotals(fdl)
```

As these are `ggplot2` objects, the output can be modified easily using 
conventional `ggplot2` syntax.

```{r}
plotReadTotals(fdl) +
    theme(
        legend.position = c(1, 0), 
        legend.justification = c(1, 0),
        legend.background = element_rect(colour = "black"))
```

### plotBaseQuals()

A heatmap of the entire `Per base sequence quality` scores can by generated, 
using the mean as representative of the scores at each position.
The median can also be selected using `plotValue = "Median"`

```{r}
plotBaseQuals(fdl)
```

The conventional boxplot for a single file can be created by specifying a 
single `FastqcData` or `FastqcFile` object.

```{r}
plotBaseQuals(fdl[[1]])
```

Boxplots of any combinations can also be drawn from a `FastqcDataList` or 
`FastqcFileList` by setting the argument `plotType = "boxplot"`.
However, this may be not suitable for datasets with numerous libraries.

```{r}
plotBaseQuals(fdl[1:4], plotType = "boxplot")
```

### plotSeqQuals()

A heatmap of mean sequence qualities can be generated, with the alternative 
being a set of overlaid lines.

```{r}
plotSeqQuals(fdl)
```

```{r}
r2 <- grepl("R2", fileName(fdl))
plotSeqQuals(fdl[r2], plotType = "line")
```

### plotSeqContent()

Heatmaps showing a composite colour for each position can be drawn. 
Colours are combined using `rgb()` with `A`,`C`, `G` and `T` being represented 
by green, blue, black and red respectively.

```{r}
plotSeqContent(fdl)
```

**NB** These plots can be very informative setting the argument 
`usePlotly = TRUE`, however they can be slow to render given the nature of how 
the package `plotly` renders graphics.

Individual plots can also be drawn using a `FastqcData` or `FastqcFile` object

```{r}
plotSeqContent(fdl[[1]])
```

Again, supplying multiple files and setting `plotType = "line"` will replicate 
the individual plots from original reports.
The argument `nc` is passed to `facet_wrap()` from the package `ggplot2`.

```{r}
plotSeqContent(fdl[1:2], plotType = "line", nc = 1)
```


### plotNContent()

N content can also be checked.
The example dataset contained no N content, and so a "blank" plot will be 
produced.

```{r}
plotNContent(fdl)
```

### plotSeqLengthDistn

In the example reports, all reads are 84bp.
Where lengths vary, a more informative heatmap will be created. 

```{r}
plotSeqLengthDistn(fdl)
```

### plotDupLevels()

Duplication levels are shown as a heatmap based on each default bin included 
in the initial FastQC reports.
By default these are the values provided in the reports as "Pre" de-duplication.

```{r}
plotDupLevels(fdl)
```

### plotAdapterContent()

Total Adapter Content is able to be plotted for complete set of files

```{r}
plotAdapterContent(fdl)
```

Adapter content by individual adapter type is available for each individual 
file, as identified by FastQC.

```{r}
plotAdapterContent(fdl[[1]]) 
```

### plotKmers()

Although not highly informative, particularly considering the binning across 
positions by FastQC, the totals of identified K-mers are plotted for a set of 
files.

```{r}
plotKmers(fdl)
```

The top 6 K-mers, as conventionally displayed by FastQC can be simply obtained, 
however K-mers which overlap at individual positions may not be visible.

```{r}
plotKmers(fdl[[1]], axis.text.x = element_text(angle = 90)) 
```

It may be worth noting that `Kmer_Content` is no longer created by default in 
later versions of FastQC.

## GC content

### The class TheoreticalGC

A selection of Theoretical GC content is supplied with the package in the object `gcTheoretical`, which has been defined with the additional `S4` class 
`TheoreticalGC`.
GC content was calculated using scripts obtained from 
https://github.com/mikelove/fastqcTheoreticalGC.
Available genomes and transcriptomes can be obtained using the functions 
`genomes()` and `transcriptomes()` on the object `gcTheoretical`

```{r, eval=FALSE}
transcriptomes(gcTheoretical)
genomes(gcTheoretical)
```

### plotGcContent()

This function offers numerous possibilities, and is able to be presented as 
the difference from a theoretical GC content.
By default, the **difference** is shown from the `Hsapiens` *genome*


```{r}
plotGcContent(fdl)
```

This can be changed to the transcriptome by setting 
`theoreticalType = "Transcriptome"`.
Alternative species can be also be selected from those made available

```{r}
plotGcContent(fdl, theoreticalType = "Transcriptome", species = "Mmusculus")
```

Customized theoretical GC content can be generated using input DNA sequences 
from a fasta file.

```{r, message=FALSE, warning=FALSE}
faFile <- system.file(
    "extdata", "Athaliana.TAIR10.tRNA.fasta", 
    package = "ngsReports")
plotGcContent(fdl, Fastafile = faFile, n = 1000)
```

Line plots can also be produced.

```{r}
plotGcContent(fdl, plotType = "line",  theoreticalType = "Transcriptome")
```

## Overrepresented Sequences

Instead of identifying common sequences across a set of libraries, 
overrepresented sequences are summarised by their possible source as defined by 
FastQC.

```{r}
plotOverrep(fdl)
```

The top `n`can be plotted for an individual file, again broken down by their 
possible source, and coloured based on their `WARN/FAIL` status

```{r, fig.wide = TRUE}
plotOverrep(fdl[[1]])
```

In addition to the above, overrepresented sequences can be exported as a FASTA 
file for easy submission to `blast`.

```{r, eval = FALSE}
exportOverrepresented(fdl, n = 10)
```

# Importing Log Files

A selection of log files as produced by tools such as `STAR`, `hisat2`, 
`bowtie`, `bowtie2` and `picard duplicationMetrics`, can be easily imported.
Files produced by `hisat2` and `bowtie2` have the identical format, and produce 
identical output

```{r}
fl <- c("bowtie2PE.txt", "bowtie2SE.txt")
bowtie2Logs <- system.file("extdata", fl, package = "ngsReports")
df <- importNgsLogs(bowtie2Logs, type = "bowtie2")
```

```{r, echo=FALSE}
pander(
    df, 
    style = "rmarkdown", 
    caption = "Example of SE and PE output from bowtie 2"
)
```

`Bowtie` provides different output to both `hisat2` and `bowtie2`

```{r}
fls <- c("bowtiePE.txt", "bowtieSE.txt")
bowtieLogs <- system.file("extdata", fls, package = "ngsReports")
df <- importNgsLogs(bowtieLogs, type = "bowtie")
```

Different information again is contained in STAR alignment logs.

```{r}
starLog <- system.file("extdata", "log.final.out", package = "ngsReports")
df <- importNgsLogs(starLog, type = "star")
```

In addition to the files produced by the above alignment tools, the output 
from Duplication Metrics (picard) can also be imported.
This is imported as a list with a `tibble` containing the detailed output in 
the list element `$metrics` and the histogram data included as the second 
elements.

```{r}
sysDir <- system.file("extdata", package = "ngsReports")
fl <- list.files(sysDir, "Dedup_metrics.txt", full.names = TRUE)
dupMetrics <- importNgsLogs(fl, type = "duplicationMetrics", which = "metrics")
dupMetrics
```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
